src/wallet/mod.rs

```diff
@@ -1689,16 +1689,75 @@ where
         let progress = progress.unwrap_or_else(|| Box::new(NoopProgress));
 
         let run_setup = self.ensure_addresses_cached(CACHE_ADDR_BATCH_SIZE)?;
-
         debug!("run_setup: {}", run_setup);
+
         // TODO: what if i generate an address first and cache some addresses?
         // TODO: we should sync if generating an address triggers a new batch to be stored
-        if run_setup {
-            maybe_await!(
-                blockchain.wallet_setup(self.database.borrow_mut().deref_mut(), progress,)
-            )?;
+
+        // first run, we record progress...
+        let mut sync_res = if run_setup {
+            maybe_await!(blockchain.wallet_setup(self.database.borrow_mut().deref_mut(), progress))
         } else {
-            maybe_await!(blockchain.wallet_sync(self.database.borrow_mut().deref_mut(), progress,))?;
+            maybe_await!(blockchain.wallet_sync(self.database.borrow_mut().deref_mut(), progress))
+        };
```

Comment:
I think you can avoid duplicating this piece of code by moving this inside the loop, and then breaking if the descriptor is not derivable.

Something like:

```rust
for _ in 0..MAX_ROUNDS {
    let sync_res = // .... ;

    if !is_derivable {
        break;
    }

    // do everything else to get the database ready for the next iteration
}
```

Reply:
I'm not sure what to do with the progress, I feel like we should have our own "internal" progress (that can be cloned and reused) that wraps the real progress and maps the first iteration to a maximum of like 70%, then if there are more iterations each has a lower "weight", like the second iteration going to 0 to 100 only causes the overall progress to go from 70 to 80 maybe and then at the end we just set it to 100.

It's not great but I can't think of any other way to predict how many iterations there's going to be. We should adjust those weights based on probability, most of the times it's just one (so that gets a larger weight).

If you have better ideas let me know.

Reply:
I did something else to avoid duplicating code, let me if you are happy with my solution :)

Regarding the `SyncProgress` object... Because right now electrum-based blockchains don't actually report progress at all, and only electrum-based blockchains will return `Error::MissingCachedScripts` (for now), I feel like we shouldn't bother changing behavior for it in this PR?

Maybe we should do another ticket to fix progress-reporting behavior. I realized that not all sync attempts will report progress (for example, rpc only reports sync progress on it's initial sync).